src/algebra/monoid_algebra/degree.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.monoid_algebra.basic
+
+/-!  #  Degree of an `add_monoid_algebra`
+
+Let `R` be a semiring and let `A` be a `semilattice_sup`.
+
+For an element `f : add_monoid_algebra R A`, this file defines
+* `add_monoid_algebra.degree`: the degree taking values in `with_bot A`,
+* `add_monoid_algebra.trailing_degree`: the trailing degree taking values in `with_top A`.
+If the grading type `A` is a linearly ordered additive monoid, then this notions of degree
+coincide with the standard one:
+* the degree is the maximum of the exponents of the monomials that appear with non-zero
+  coefficient in `f`, or `⊥`, if `f = 0`;
+* the trailing degree is the minimum of the exponents of the monomials that appear with non-zero
+  coefficient in `f`, or `⊤`, if `f = 0`.
+
+Currently, the only results are
+* `degree_mul_le` -- the degree of a product is at most the sum of the degrees,
+*  `le_trailing_degree_mul` -- the trailing degree of a product is at least the sum of the
+  trailing degrees.
+-/
+
+namespace add_monoid_algebra
+open_locale classical
+
+variables {R A B : Type*} [semiring R]
+
+section general_results_assuming_semilattice_sup
+
+variables [has_add A]
+variables [semilattice_sup B] [order_bot B] [has_add B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+
+lemma sup_support_mul_le {D : A → B} (Dm : ∀ {a b}, D (a + b) ≤ D a + D b)
+  (f g : add_monoid_algebra R A) :
+  (f * g).support.sup D ≤ f.support.sup D + g.support.sup D :=
+begin
+  refine (finset.sup_le (λ d ds, _)),
+  obtain ⟨a, af, b, bg, rfl⟩ : ∃ a, a ∈ f.support ∧ ∃ b, b ∈ g.support ∧ d = a + b,
+  { simpa only [finset.mem_bUnion, finset.mem_singleton, exists_prop] using f.support_mul g ds },
+  refine Dm.trans (add_le_add _ _);
+  exact finset.le_sup ‹_›,
+end
```

Comment:
Can you add the following three lemmas too?

* `sup_support_list_prod_le`
* `sup_support_multiset_prod_le`
* `sup_support_finset_prod_le`
* `sup_support_pow_le`

The first should follow by induction, the next two by definition, and the last by expanding the power into a sum

Reply:
I added these lemmas: I hope that my statements were what you had in mind!

The flurry of assumptions is due to empty lists/multisets/finsets.  Probably with `fold` they could be simplified, but I would rather not do that!